pkg/virt-operator/install-strategy/create.go

```diff
@@ -0,0 +1,995 @@
+				return false, fmt.Errorf("unable to create crd %+v: %v", crd, err)
+			}
+			log.Log.V(2).Infof("crd %v created", crd.GetName())
+
+		} else if !objectMatchesVersion(&cachedCrd.ObjectMeta, imageTag, imageRegistry) {
+			// Patch if old version
+			var ops []string
+
+			if crd.Spec.Version != cachedCrd.Spec.Version {
+				// We can't support transitioning between versions until
+				// the conversion webhook is supported.
+				return false, fmt.Errorf("No supported update path from crd %s version %s to version %s", crd.Name, cachedCrd.Spec.Version, crd.Spec.Version)
+			}
+
+			// Add Labels and Annotations Patches
+			labelAnnotationPatch, err := createLabelsAndAnnotationsPatch(&crd.ObjectMeta)
+			if err != nil {
+				return false, err
+			}
+			ops = append(ops, labelAnnotationPatch...)
+
+			// Add Spec Patch
+			newSpec, err := json.Marshal(crd.Spec)
+			if err != nil {
+				return false, err
+			}
+			ops = append(ops, fmt.Sprintf(`{ "op": "replace", "path": "/spec", "value": %s }`, string(newSpec)))
+
+			_, err = ext.ApiextensionsV1beta1().CustomResourceDefinitions().Patch(crd.Name, types.JSONPatchType, generatePatchBytes(ops))
+			if err != nil {
+				return false, fmt.Errorf("unable to patch crd %+v: %v", crd, err)
+			}
+			log.Log.V(2).Infof("crd %v updated", crd.GetName())
+
+		} else {
+			log.Log.V(4).Infof("crd %v is up-to-date", crd.GetName())
+		}
+	}
+
+	// create/update ServiceAccounts
+	for _, sa := range targetStrategy.serviceAccounts {
+		var cachedSa *corev1.ServiceAccount
+
+		sa := sa.DeepCopy()
+		obj, exists, _ := stores.ServiceAccountCache.Get(sa)
+		if exists {
+			cachedSa = obj.(*corev1.ServiceAccount)
+		}
+
+		injectOperatorLabelAndAnnotations(&sa.ObjectMeta, imageTag, imageRegistry)
+		if !exists {
+			// Create non existent
+			expectations.ServiceAccount.RaiseExpectations(kvkey, 1, 0)
+			_, err := core.ServiceAccounts(kv.Namespace).Create(sa)
+			if err != nil {
+				expectations.ServiceAccount.LowerExpectations(kvkey, 1, 0)
+				return false, fmt.Errorf("unable to create serviceaccount %+v: %v", sa, err)
+			}
+			log.Log.V(2).Infof("serviceaccount %v created", sa.GetName())
+		} else if !objectMatchesVersion(&cachedSa.ObjectMeta, imageTag, imageRegistry) {
+			// Patch if old version
+			var ops []string
+
+			// Patch Labels and Annotations
+			labelAnnotationPatch, err := createLabelsAndAnnotationsPatch(&sa.ObjectMeta)
+			if err != nil {
+				return false, err
+			}
+			ops = append(ops, labelAnnotationPatch...)
+
+			_, err = core.ServiceAccounts(kv.Namespace).Patch(sa.Name, types.JSONPatchType, generatePatchBytes(ops))
+			if err != nil {
+				return false, fmt.Errorf("unable to patch serviceaccount %+v: %v", sa, err)
+			}
+			log.Log.V(2).Infof("serviceaccount %v updated", sa.GetName())
+
+		} else {
+			// Up to date
+			log.Log.V(2).Infof("serviceaccount %v already exists and is up-to-date", sa.GetName())
+		}
+	}
+
+	// create/update ClusterRoles
```

Comment:
hm, updade is tricky here, especially if privileges are removed. I think we first have to detect the difference, add new permissions and after cluster-update remove the old ones. One way to do this, is to have them e.g. in parallel (so during the update allow more), and then remove the old cluster role afterwards.

Reply:
i agree. I think this pattern is complex enough to justify a follow up PR with a dedicated discussion. Since this impacts us once only once permissions change, we have at least one release cycle to work through this.